src/zenml/integrations/kubernetes/orchestrators/dag_runner.py

```diff
@@ -0,0 +1,181 @@
+#  Copyright (c) ZenML GmbH 2022. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""DAG (Directed Acyclic Graph) Runners."""
+
+import threading
+from collections import defaultdict
+from typing import Any, Callable, Dict, List
+
+from zenml.logger import get_logger
+
+logger = get_logger(__name__)
+
+
+def reverse_dag(dag: Dict[str, List[str]]) -> Dict[str, List[str]]:
+    """Reverse a DAG.
+
+    Args:
+        dag: Adjacency list representation of a DAG.
+
+    Returns:
+        Adjacency list representation of the reversed DAG.
+    """
+    reversed_dag = defaultdict(list)
+
+    # Reverse all edges in the graph.
+    for node, upstream_nodes in dag.items():
+        for upstream_node in upstream_nodes:
+            reversed_dag[upstream_node].append(node)
+
+    # Add nodes without incoming edges back in.
+    for node in dag:
+        if node not in reversed_dag:
+            reversed_dag[node] = []
+
+    return reversed_dag
+
+
+class ThreadedDagRunner:
+    """Multi-threaded DAG Runner."""
+
+    def __init__(
+        self, dag: Dict[str, List[str]], run_fn: Callable[[str], Any]
+    ) -> None:
+        """Init.
+
+        Args:
+            dag: Adjacency list representation of a DAG.
+                E.g.: [(1->2), (1->3), (2->4), (3->4)] should be represented as
+                `dag={2: [1], 3: [1], 4: [2, 3]}`
+            run_fn: A function `run_fn(node)` that runs a single node
+        """
+        self.dag = dag
+        self.reversed_dag = reverse_dag(dag)
+        self.run_fn = run_fn
+        self.nodes = dag.keys()
+        self.node_is_waiting = {node: True for node in self.nodes}
+        self.node_is_running = {node: False for node in self.nodes}
+        self.node_is_completed = {node: False for node in self.nodes}
```

Comment:
Super nitpicky and not necessary to fix at all, but I think this would be a little cleaner with a `State` enum or something instead of three dicts:
```python
self.node_states = {node: SomeGoodEnumName.WAITING for node in self.nodes }
```

Reply:
Changed it to use a Enum instead.